File: src/sync_mdx.py
import os
import re
import datetime
import json
from src.paths import generated_root
from src.publishers.notion import query_published_pages, retrieve_page_markdown

def safe_slug(markdown: str, page_id: str) -> str:
    title_match = re.search(r"^#\s+(.+)$", markdown, re.MULTILINE)
    title = title_match.group(1) if title_match else page_id
    
    cleaned = title.strip().lower()
    slug = re.sub(r"[^\w\s-]", "", cleaned, flags=re.UNICODE)
    slug = re.sub(r"[\s_]+", "-", slug)
    slug = re.sub(r"-+", "-", slug)
    return slug if slug else page_id.replace("-", "")

def quote_yaml(value: str) -> str:
    return json.dumps(value, ensure_ascii=False)

def sync_mdx() -> int:
    os.makedirs(generated_root, exist_ok=True)
    pages = query_published_pages()
    
    for page in pages:
        page_id = page.get("id")
        markdown = retrieve_page_markdown(page_id)
        
        title_match = re.search(r"^#\s+(.+)$", markdown, re.MULTILINE)
        title = title_match.group(1) if title_match else page_id
        
        slug = safe_slug(markdown, page_id)
        
        iso_now = datetime.datetime.utcnow().isoformat() + "Z"
        output = [
            "---",
            f"title: {quote_yaml(title)}",
            f"slug: {quote_yaml(slug)}",
            f"notionPageId: {quote_yaml(page_id)}",
            f"syncedAt: {quote_yaml(iso_now)}",
            "---",
            "",
            markdown,
            ""
        ]
        
        file_path = generated_root / f"{slug}.mdx"
        with open(file_path, "w", encoding="utf-8") as f:
            f.write("\n".join(output))
            
    return len(pages)

File: src/sync-mdx.ts
import { mkdir, writeFile } from "node:fs/promises";
import path from "node:path";
import { generatedRoot } from "./paths.js";
import { queryPublishedPages, retrievePageMarkdown } from "./publishers/notion.js";

function safeSlug(markdown: string, pageId: string): string {
  const title = markdown.match(/^#\s+(.+)$/m)?.[1] ?? pageId;
  const slug = title
    .normalize("NFKD")
    .replace(/[^\w\s-]/g, "")
    .trim()
    .toLowerCase()
    .replace(/[\s_]+/g, "-");
  return slug || pageId.replaceAll("-", "");
}

function quoteYaml(value: string): string {
  return JSON.stringify(value);
}

export async function syncMdx(): Promise<number> {
  await mkdir(generatedRoot, { recursive: true });
  const pages = await queryPublishedPages();

  for (const page of pages) {
    const markdown = await retrievePageMarkdown(page.id);
    const title = markdown.match(/^#\s+(.+)$/m)?.[1] ?? page.id;
    const slug = safeSlug(markdown, page.id);
    const output = [
      "---",
      `title: ${quoteYaml(title)}`,
      `slug: ${quoteYaml(slug)}`,
      `notionPageId: ${quoteYaml(page.id)}`,
      `syncedAt: ${quoteYaml(new Date().toISOString())}`,
      "---",
      "",
      markdown,
      ""
    ].join("\n");
    await writeFile(path.join(generatedRoot, `${slug}.mdx`), output, "utf8");
  }
  return pages.length;
}

